### PR TITLE
Improvements to concurrent tasks

### LINKED ISSUE #157: Show error on failed version refresh
### Role

I use minecraft

### Suggestion

Prism Launcher should show a GUI error when fetching version info (and other remote resources) fails.

### Benefit

Much easier troubleshooting of version problems.

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

Prism Launcher was only showing versions of Minecraft up to 1.19, despite 1.19.2 being out at time of writing. Refreshing seemed to go through but did not fix the issue. Checking the console, I saw the following output:
```
    4.470 D Task "NetJob(Download of meta file net.minecraft/index.json)" starting for the first time
    4.479 D Task "Net::Download(0x55e0200ada40)" starting for the first time
    4.479 D Downloading  "https://meta.scrumplex.rocks/net.minecraft/index.json"
    5.340 C Failed  "https://meta.scrumplex.rocks/net.minecraft/index.json"  with reason  QNetworkReply::ContentNotFoundError
    5.340 D Download failed in previous step: "https://meta.scrumplex.rocks/net.minecraft/index.json"
    5.341 D Task "Net::Download(0x55e0200ada40)" restarting for after failing at first
    5.341 D Downloading  "https://meta.scrumplex.rocks/net.minecraft/index.json"
    5.698 C Failed  "https://meta.scrumplex.rocks/net.minecraft/index.json"  with reason  QNetworkReply::ContentNotFoundError
    5.698 D Download failed in previous step: "https://meta.scrumplex.rocks/net.minecraft/index.json"
    5.698 D Task "Net::Download(0x55e0200ada40)" restarting for after failing at first
    5.698 D Downloading  "https://meta.scrumplex.rocks/net.minecraft/index.json"
    5.874 C Failed  "https://meta.scrumplex.rocks/net.minecraft/index.json"  with reason  QNetworkReply::ContentNotFoundError
    5.875 D Download failed in previous step: "https://meta.scrumplex.rocks/net.minecraft/index.json"
```
This is because the old recommended meta server was shut down, which is also something that should be fixed (CNAME or redirect it to the new one).
However, this error was completely invisible to the user in the GUI, and just made it seem like no new version had been released. On top of fixing the old meta server, these errors should definitely be shown to the user in the GUI so that they know something has gone wrong, and can understand why.

### LINKED ISSUE #659: Prism Launcher can crash when clicking OK button in update mods screen
### Operating System
Windows 11

### Version of Prism Launcher

Prism Launcher 6.1

### Version of Qt

Qt 6.4.0

### Description of bug

When attempting to update mods, Prism may crash. I clicked the check for updates button in the mods section for an instance. There were no updates, but upon clicking OK on the "all mods are up to date" box, Prism crashes.

I wish I could reliably reproduce it, but it seems to happen at random when clicking that OK button.

Additionally, this crash also happened yesterday where there *was* an update to a mod. I said yes to the update and after the launcher still crashed (the update was applied correctly however).

I do not see any indication of a crash being logged in any of the PrismLauncher-<number>.log files, but I do see the crash in the Windows Reliability Monitor (and Event Viewer). I'm not sure how much this would help, but here is an image of what I see there: https://i.imgur.com/h2yS4MQ.png. Looks to have _something_ to do with Qt (although that probably doesn't narrow things down).

The closest I saw to an error is 
` 7.742 W QObject::connect: No such signal QNetworkReplyHttpImpl::error(QNetworkReply::NetworkError)
`,
but I don't know if this would cause a crash like this. 

### Steps to reproduce

- Click an instance 
- Click the mods tab on the side 
- Click check for updates
- The launcher may or may not crash after clicking the OK button in the box saying that all mods are up to date

### Suspected cause

N/A

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #664: Launcher freezes/crashes after scrolling through Modrinth modpacks (Linux)
### Operating System

Linux

### Version of Prism Launcher

Prism Launcher 6.1

### Version of Qt

Qt 5.15.7

### Description of bug

When adding a new instance and scrolling through the list of Modrinth modpacks, the launcher freezes up and segfaults. It happens consistently when using Linux, though I couldn't replicate it on Windows 10. No other mod provider has this issue.

https://user-images.githubusercontent.com/82874329/209215184-84c06706-bcde-491f-a975-921073737269.mp4

### Steps to reproduce

* Click "Add Instance"
* Select Modrinth
* Scroll the modpack list as far as possible

### Suspected cause

_No response_

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #689: Updating already updated instances crashes the launcher
### Operating System

Windows

### Version of Prism Launcher

Prism Launcher 6.1

### Version of Qt

Qt 6.4.1

### Description of bug

Clicking 'Update Instance' in the CurseForge tab when the instance is already updated crashes the launcher

### Steps to reproduce

1. Download modpack of choice thru the launcher
2. Click on the CurseForge tab of that modpack
3. Update the instance to the latest version

### Suspected cause

_No response_

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #863: Quick scrolling of the modpack list at adding instance screen crashes the app on windows 10
### Operating System

Windows

### Version of Prism Launcher

Prism Launcher 6.3

### Version of Qt

Qt 6.4.2

### Description of bug

When in new instance screen and want to get a new modpack. When you scroll quickly the app crashes!

### Steps to reproduce

* Go to Adding new instance menu
* Select Modrinth/CurseForge
* Scroll down with mouse quickly
* Crash

### Suspected cause

It only occurs when the images doesn't load yet so I think it could be that app is trying to access the value of a page that doesn't exist so app is throwing an range error.
On the other hand I don't know, because there is no error message after the crash.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #1340: Crashes when I exit after aborting mod update
### Operating System

macOS

### Version of Prism Launcher

Prism Launcher 7.1

### Version of Qt

Qt 6.5.0

### Description of bug

When I exit the mod screen after aborting an update the whole thing crashes

### Steps to reproduce

1. Go to mods
2. update mods
3. abort mod update
4. Close the mod edit window
5. And it crashes

### Suspected cause

Mod updater was a bit slow. Maybe it's that?

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #1430: Prismlauncher segfaults when updating instance
### Operating System

Arch Linux 64-bits

### Version of Prism Launcher

Prism Launcher 7.1
Commit: c264d0ad5942df9523fe1d52357f392e2a904350

### Version of Qt

Qt 5.15.10

### Description of bug

It's only after getting a bit into the update, it doesn't happen instantaneously. It also appears to happen at an arbitrary step in the process. Worst of all, it doesn't segfault every single time, just about 9/10 times and I end up just having to restart the launcher and try to update over and over again until it decides not to segfault.

I've experienced this issue since the prismlauncher fork.

[GDB backtrace](https://gist.github.com/dhasial/e790ab259e422c6a5245c0c14a48eb8d)

### Steps to reproduce

 Click on the "Curseforge" tab in the instance, then click the "update pack" button.

### Suspected cause

_No response_

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #1638: [dev build] Crash on mass mod update
### Operating System

Windows

### Version of Prism Launcher

8.0-develop (latest)
Commit: 98bc102f5bec14b38c684c219989fc3953dc2f49

### Version of Qt

Qt 6.5.2

### Description of bug

I believe this relies on an instance with a large amount of mods, and a lot of mods with updates.

Reading the log file, there is a very large chunk filled with `NULL`, which makes it very difficult to post in here or mclo.gs

### Steps to reproduce

- get an instance with 200+ mods
- have a large amount of them awaiting an update (?)
- hit "update" on unselected or after CTRL-A
- proceed with the update request
- wait for the updates to complete

*- and Prism crashes

### Suspected cause

A stack overflow or data overload of some kind. It is possible that I ran into this bug by chance and the causes I have cited are not related.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #1899: Crash when searching by author in Curseforge tab by Copy Paste Then Manually Searching for Another Author (Finicky bug)
### Operating System

_No response_

### Version of Prism Launcher

8.0 Stable

### Version of Qt

6.6.0

### Description of bug

When searching for an author by copy pasting a username and then afterword manually typing a user name and click on a mod it crashes. 

### Steps to reproduce

Originally it was copy and paste a author username and then manually write in a different author.

### Suspected cause

Not sure. https://gist.github.com/P3rf3ctXZer0/0603f5e3aecc655ea6d4def26a9e2569

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### LINKED ISSUE #1919: Installing a CurseForge modpack can silently fail to download some mods.
### Operating System

Windows

### Version of Prism Launcher

Prism 8.0

### Version of Qt

Qt 6.6.0

### Description of bug

Prism failed to download certain mods and did not give any obvious indication or interruption that it failed to download them.
Not to be confused with the mods which it could not download because of CurseForge, which I manually downloaded and it located, which it complained loudly about as is correct for it to do.

### Steps to reproduce

1. Attempt to install the "All Of Fabric 6" pack from CurseForge
2. Experience a network transient(?) either clientside or serverside at the exact moment that Prism attempts to download mods from Modrinth instead of CurseForge.

### Suspected cause

So for that particular pack and the current arrangement of everything, there will be mods from 3 sources. CurseForge api, CurseForge manually, and Modrinth.
`   218.424 C | [launcher.task.net.download]: "{47c5c89e-3171-41bf-bb05-26d4aa143553}" Failed  "https://cdn.modrinth.com/data/2g6vOSXZ/versions/MfYeoZyD/fishing_ruler-1.0.1-mc1.19%2B.jar"  with reason  QNetworkReply::ServiceUnavailableError`
`   218.426 D | Setting up api download`
`   218.519 D | Setting up api download`
`   218.530 C | [launcher.task.net.download]: "{1d274a55-1809-45d2-8460-4c35d22676f5}" Failed  "https://cdn.modrinth.com/data/hI3Tmydh/versions/lzVTnJPm/muffins-slimegolem-1.19.x-1.1.1-fabric.jar"  with reason  QNetworkReply::ServiceUnavailableError`
Prism correctly infers that some of the mods which would have to be downloaded manually from CurseForge could be downloaded automatically from Modrinth, and attempts to set that up.
It seems that at the exact moment it actually tried to download them, it just couldn't.

### This issue is unique

- [X] I have searched the issue tracker and did not find an issue describing my bug.

### PR DESCRIPTION
**Pull request by Trial97**
I did get a crash:
- add a new instance 
- technic
- Attack for B team latest version: https://www.technicpack.net/modpack/attack-of-the-bteam.552556 
- click OK
- after it starts to download mods click on abort
- I get a crash

Env:
- latest develop (a7bdfb5f6b43b1300e4a081ee358e898a118643e)
- Qt 6.5.3
- Number of concurrent tasks 24
- Number of concurrent downloads 6

I have high hopes that this may solve some other issues regarding concurrent tasks

Here are some of the fixed issues:
- fixes #1638 (I was able to reproduce it and fix it)
- fixes #659 (same as above)
- fixes #1430(same as above)
- fixes #689
- fixes #1340 (this is the same issue described in this PR)
- fixes #863
- fixes #664
- fixes #157 (we already have the messagebox but NetJob did not ever fail; prior to this PR)
- fixes #1899 (https://github.com/PrismLauncher/PrismLauncher/pull/1694/commits/2d91bd09d413177c9382e7b101cc4cd8f7d78b45)
- fixes https://github.com/PrismLauncher/PrismLauncher/pull/1263#issuecomment-1610113772 on https://github.com/PrismLauncher/PrismLauncher/pull/1694/commits/463ff4c6ae9362c70e4f70eea0805ce7b6931145
- fixes #1919 (by the fact that now concurrent task fails if one task fails) 

Once this is released we should notify:
- https://github.com/PrismLauncher/PrismLauncher/issues/1299
- https://github.com/PrismLauncher/PrismLauncher/issues/1205
- https://github.com/PrismLauncher/PrismLauncher/issues/1138
- https://github.com/PrismLauncher/PrismLauncher/issues/936

### CODE REVIEW

File: launcher/tasks/ConcurrentTask.cpp [hunk 1]
```diff
@@ -108,27 +109,33 @@ void ConcurrentTask::clear()
...
-
-    if (m_queue.isEmpty() && m_doing.isEmpty() && !wasSuccessful()) {
-        emitSucceeded();
+    }
+    if (m_queue.isEmpty()) {
+        if (m_doing.isEmpty()) {
+            // if (m_failed.isEmpty())
+            emitSucceeded();
+            // else
+            // emitFailed(tr("One or more subtasks failed"));
```
Comment (Trial97) [inline launcher/tasks/ConcurrentTask.cpp]: I'm considering merging this PR with https://github.com/PrismLauncher/PrismLauncher/pull/1364 as both modify the ConncurentTask in a way that creates a need for heavy testing

**Comment by Trial97** [comment]
Removed from milestone until I have a clear idea for this

File: launcher/tasks/ConcurrentTask.cpp (hunk 1 above)
Comment (Trial97) [reply launcher/tasks/ConcurrentTask.cpp]: decided to merge the two PRs.

**Comment by Trial97** [comment]
This is the successor of https://github.com/PrismLauncher/PrismLauncher/pull/1364.
Right now this is not stable (tests may still fail on actions).
I need to test most tasks(similar to the old PR)

**Comment by Trial97** [comment]
Now that I magically solved the action tests, here is a list with all the stuff I will test in order to see if this works or not:(with asan)

# ConcurrentTask
- [x] Create a Custom instance
- [x] Create an Import instance:
  - [x] from URL
  - [x] from Curseforge URL
  - [x] Curseforge
  - [x] Modrinth
  - [x] Prism
- [x] Create an AtLauncher instance
- [x] Create a Curseforge instance 
  - [x] with blocked mods
- [x] Create a Ftb Legacy instance:
  - [x] public
  - [x] Third Party
  - [x] Private
- [x] Create a Ftb App Import instance
- [x] Create a Modrinth instance
- [x] Create a Technic instance
- [x] try to abort the task when it's running for the add instance action
- [x] Update mods:
  - [x] with metadata
  - [x] without metadata
  - [x] dependency
- [x] Export pack:
  - [x] Prism
  - [x] Modrinth
  - [x] Curseforge
- [x] Download:
  - [x] mods
  - [x] mods with dependency
  - [x] resource packs
  - [x] shaders
  - [x] textures
- [x] That resources are correctly displayed(mods/resource packs/shaders/textures)
# SequentialTask
- [x] Imgur Uplad with album
- [x] Skin upload
- [x] update mods
- [x] check for mods dependencies
# NetJob 
- [x] all download dialogs:
  - [x] Curseforge
  - [x] Modrinth
- [x] all mod packs
- [x] the mod/mod pack/resource description is loaded correctly
- [x] the meta (versions and loaders)
- [x] the asset download (the stuff that is prompted right before you play the game)
- [x] all loaders work(with libraries and stuff):
  - [x] vanilla
  - [x] forge
  - [x] neoforge
  - [x] fabric
  - [x] quilt 
  - [x] litleloader
- [x] the online resource logo loads correctly
- [x] the news is loaded
- [x] the translation is loading
- [x] the curseforge modpack import works (from website)
- [x] the curseforge mod import works (from website)
- [x] managed pack page works:
  - [x] Curseforge
  - [x] Modrinth
  - [x] and version change for both providers
  - [x] and version refresh for both providers
- [x] Imgur upload 
# extra
- [x] about dialog
- [x] account list page works
- [x] try to move through the resource lists rapidly
# MultipleOptionsTask is not used so there is no test for it


** This list is not final, it's just a list from a quick search through code for ConccurentTask stuff. If I missed any test case let me know.

* I will reset this list if I make changes to the PR(meaning I need to test all cases again)

**Comment by Trial97** [comment]
Tested all that complicated stuff and it all works now.
With that, this is ready to review and test.
As I modified the ConncurentTask stuff I expect at least one other person(or more) will test this PR. 
And a small question for somebody with more qt knowledge:
If I define a QSharedPointer, then connect a signal from that pointer to lambda and capture sharedPointer through value; Will that pointer be deleted automatically, or do I need some code that calls deleteLater or disconnect on that pointer?

**Comment by Trial97** [comment]
Now this was tested with asan enabled so we should be good on the memory side

**Review by ColonelGerdauf — APPROVED** [review]

**Review by TayouVR — APPROVED** [review]
I am not super confident in my testing, but I was not able to create a crash or otherwise bad behaviour... so it looks good to me.
More people please test!

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**